### ISSUE TITLE
Flipped W^T W and W W^T operation

### ISSUE THREAD

**Issue by Marius-Juston** [araujoalexandre/Lipschitz-SLL-Networks#3]
Hello,

I noticed that in the paper, you have it annotated that the matrix T should be computed as 

$T_{ii} = \sum^n_{j=1}\left|W^TW\right|_{ij} \frac{q_j}{q_i}$

`t = torch.abs(torch.einsum('i,ik,kj,j -> ij', q_inv, self.weight.T, self.weight, q)).sum(1)`

while in your implementation you have it so that

`t = torch.abs(torch.einsum('i,ik,kj,j -> ij', q_inv, self.weight, self.weight.T, q)).sum(1)` 

which would be equivalent to 

$T_{ii} = \sum^n_{j=1}\left|WW^T\right|_{ij} \frac{q_j}{q_i}$

is this an error?

**Comment by Marius-Juston** [comment]
As a quick check, we can look at the eigenvalues, since:

$W^T W \preceq T$ needs to be true.

```python

import torch


class Distribution:
    def __init__(self, N=10):
        # self.weight = torch.randn((N, N))
        self.weight = torch.arange(N ** 2).reshape((N, N)).to(dtype=torch.float)
        self.q = torch.ones(N)

    def compute_t(self):
        q = self.q
        q_inv = torch.reciprocal(self.q)
        t = torch.abs(torch.einsum('i,ik,kj,j -> ij', q_inv, self.weight, self.weight.T, q)).sum(1)
        return t


def check_validity(W, T):
    T = torch.diag(T)

    outputs = torch.linalg.eigvals(W.T @ W - T)

    negs = outputs.real <= 0

    print(torch.round(outputs.real, decimals=2))

    return torch.all(negs)


if __name__ == '__main__':
    N = 10

    dist = Distribution(N)

    W = dist.weight

    Tcorrect = torch.abs(W.T @ W).sum(1)
    print(check_validity(W, Tcorrect ))

    Tcurrent = dist.compute_t()
    print(check_validity(W, Tcurrent ))
```

The Tcurrent is what is used in the code for the SDPBasedLipschitzResidualLinearLayer file, the Tcorrect is what it should be.

For Tcorrect the eigenvalues are:

[  0.0000, -306624.3125, -348286.1250, -311896.3125, -342977.4375, -317091.5938, -337762.0938, -332583.0625, -322260.0938, -327419.6875]

For the Tcurrent the eigenvalues are:

[131358.0156,  -43586.8086, -457128.4688, -405247.9062,  -97112.6094, -353969.4375, -302877.4688, -149160.5312, -251804.2812, -200620.5000]

It can be noticed that for Tcurrent not all eigenvalues are negative semi-definitive eigenvalues.

**Comment by araujoalexandre** [comment]
Both notation are correct, you would just need to define the correct shape in the network such that $W^T W$ or $W W^T$ is of shape $n \times n$ where $n$ is the dim of x. See footnote 1 in the paper page 3. Of course, you also need to adjust the computation of $T$ accordingly. 

Don't hesitate if you have more questions!

**Comment by Marius-Juston** [comment]
Thank you for the quick response!

Sorry, I am still confused how the statements are equivalent for $W^TW$ and $WW^T$. If the matrix initialization of $W$ is not a symmetric matrix. 
In the example above, the matrix W is a square matrix, so it would be returning a shape of $n \times n$ at all times. However, it does not return all semi-negative eigenvalues for the $WW^T$ case. How would the computation of $T$ need to change between the two methods then?  For the case where it is $W^T W$ would you need to sum over the 0'th dimension then?

**Comment by Marius-Juston** [comment]
I believe that the reference `Orthogonalizing Convolutional Layers with the Cayley Transform` for footnote 1 in the paper, page 3, initializes the matrix such that the Cayley transform of the skew-symmetric matrix is always orthogonal. Where in this case, $AA^T = A^TA$ (at least from my quick reading), in the code, I did not see this constraint being applied to the weights

**Comment by araujoalexandre** [comment]
No, the paper of the repo : https://arxiv.org/pdf/2303.03169

it’s just a change of variable.

**Comment by Marius-Juston** [comment]
Oh, okay, that makes more sense. I had not realized that instead of doing $W^T x + b$ in the code you had done $Wx + b$ as well! Thanks!
